### PR TITLE
Added basic yamux connection upgrade

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

Yamux provides multiplexing over a ordered reliable connection (e.g. TCP).
More info: https://github.com/hashicorp/yamux/blob/master/spec.md

- Added yamux as a git submodule. Run `git submodule init` and `git submodule update --recursive --remote`. Optionally, run` git config submodule.recurse true` to enable updating all submodules on pull. 
- This is working from a fork of yamux (develop branch), where the
  upgrade to futures 0.3.x is close to complete
- Made a TcpStream wrapper struct which implements futures `AsyncWrite` and `AsyncRead` to reduce the tie in to tokio
- _Side note:_ investigated upgrading to futures 0.3.x however there ended up being a few external libraries which are locked to alpha futures. This is WIP for tonic (https://github.com/hyperium/tonic/pull/163) and tower (no PR - we may be able to pretty easily remove this dependency)    

## Motivation and Context

Partial #1049 

## How Has This Been Tested?

Unit tests

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: comms/Cargo.toml [hunk 1]
```diff
@@ -35,6 +35,8 @@ time = "0.1.42"
 tokio = "0.2.0-alpha.6"
 tokio-executor = { version ="^0.2.0-alpha.6", features = ["threadpool"] }
 ttl_cache = "0.5.1"
+# TODO: For now, perhaps release this as a crate `tari-yamux` until yamux 0.3 is released (requires upgrade to futures 0.3.x)
```
Comment (sdbondi) [inline comms/Cargo.toml]: Will add a submodule to git on subsequent PRs

**Comment by CjS77** [comment]
Looks good; but would you mind adding your HOWTO to the README file?

People won't know to run `git init submodules ...` and will moan when `cargo build` fails

File: README.md [hunk 2]
```diff
@@ -35,6 +36,18 @@ to generate the documentation. The generated html sits in `target/doc/`. Alterna
 
 See [RFC-0110/CodeStructure](./RFC/src/RFC-0010_CodeStructure.md) for details on the code structure and layout.
 
+## Git submodules
```
Comment (sdbondi) [inline README.md]: @CjS77 Added this section on submodules

File: README.md [hunk 3]
```diff
@@ -35,6 +35,18 @@ to generate the documentation. The generated html sits in `target/doc/`. Alterna
 
 See [RFC-0110/CodeStructure](./RFC/src/RFC-0010_CodeStructure.md) for details on the code structure and layout.
 
+### Git submodules
```
Comment (sdbondi) [inline README.md]: @CjS77 Added this section

File: README.md [hunk 4]
```diff
@@ -35,6 +35,20 @@ to generate the documentation. The generated html sits in `target/doc/`. Alterna
...
```
Comment (sdbondi) [inline README.md]: @CjS77 Added this section

**Pull request merged**